### PR TITLE
feat(deployment) Allow overriding service host and ports with env variable

### PR DESCRIPTION
**Pull request by ViViDboarder**
This won't change what host and port the service binds to, but changes the host and port used by other services to reach each other. This should allow running on systems like Kubernetes and Nomad.

This should address #647

A question for the maintainers, I can see that `immich-microservices` binds to port `3002`, however it doesn't look like any of the other services call that port. Is there somewhere I should be updating this as well?

I've been unable to verify this works yet because when I run `make dev` on the `main` branch I get this error:

```
immich-web_1               | Error: EACCES: permission denied, mkdir '/usr/src/app/node_modules/.vite/deps_temp'
immich-web_1               |     at Object.mkdirSync (node:fs:1349:3)
immich-web_1               |     at runOptimizeDeps (file:///usr/src/app/node_modules/vite/dist/node/chunks/dep-a713b95d.js:42258:14)
immich-web_1               |     at Timeout._onTimeout (file:///usr/src/app/node_modules/vite/dist/node/chunks/dep-a713b95d.js:41675:54)
immich-web_1               |     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
immich-web_1               |   errno: -13,
immich-web_1               |   syscall: 'mkdir',
immich-web_1               |   code: 'EACCES',
immich-web_1               |   path: '/usr/src/app/node_modules/.vite/deps_temp'
immich-web_1               | }
```

So I'll be debugging that.

### CODE REVIEW

**Comment by alextran1502** [comment]
Are you using the same dev environment as the production environment, or is this a fresh instance? I know that if you already running an instance in the production environment if you run dev on that instance, it might cause some issues.

Can you add the new environment variable to the `.env.example` file?

**Comment by ViViDboarder** [comment]
No. I had never run before, so it’s fresh. I’ll update the example file.

**Comment by ViViDboarder** [comment]
The issue may have been a permissions issue with Docker running in Colima on macOS. I tried on my Linux machine and running the `main` branch worked. Running my branch shows no errors in the log, but when I load the web proxy I see:

```
Error code 500 
Internal Error 
Verbose
{"message":"Internal Error"}
```

So this shows in the browser, but nothing in the logs.

**Comment by alextran1502** [comment]
> The issue may have been a permissions issue with Docker running in Colima on macOS. I tried on my Linux machine and running the `main` branch worked. Running my branch shows no errors in the log, but when I load the web proxy I see:
> 
> ```
> Error code 500 
> Internal Error 
> Verbose
> {"message":"Internal Error"}
> ```
> 
> So this shows in the browser, but nothing in the logs.

This usually means the server has a problem so that the web application cannot communicate to the server.

**Comment by ViViDboarder** [comment]
Ok, I'm stepping through my changes one at a time now. I fixed an issue with the nginx cmd and now trying to fix an issue with `immich-web`.

It's very strange to me, but I'm not very experienced in nodejs.

In order to really narrow things down, I've made changes which should be noops to compare. This runs just fine and logs two identical URLs:

```javascript
const immich_server_host = 'immich-server';
const immich_server_port = 3001;
var immich_server_url = 'http://' + immich_server_host + ':' + immich_server_port;
console.log('Looking for immich-server at url: ' + immich_server_url);
console.log('Original    immich-server at url: ' + 'http://immich-server:3001');
immich_server_url = 'http://immich-server:3001'
serverApi.setBaseUrl(immich_server_url);
```

While this shows the error I described above but logs the same two identical URLs.

```javascript
const immich_server_host = process.env.IMMICH_SERVER_HOST || 'immich-server';
const immich_server_port = process.env.IMMICH_SERVER_PORT || 3001;
var immich_server_url = 'http://' + immich_server_host + ':' + immich_server_port;
console.log('Looking for immich-server at url: ' + immich_server_url);
console.log('Original    immich-server at url: ' + 'http://immich-server:3001');
immich_server_url = 'http://immich-server:3001'
serverApi.setBaseUrl(immich_server_url);
```

In either case, they are both using the original hardcoded URL in the `setBaseUrl` method and the values of `immich_server_host` and `immich_server_port` are ignored for all but the log statement. It also doesn't appear to be a server error in web as I see no stack trace and the log calls work just fine.

**Comment by ViViDboarder** [comment]
It looks to me like, for some reason, just using `process.env.*` in any way "breaks" the server, even if it doesn't raise an error (continues to the next line) or use the returned value.

**Comment by jrasm91** [comment]
Might have something to do with this line (2nd to last line) after the print statements.
```js
immich_server_url = 'http://immich-server:3001'
```

**Comment by ViViDboarder** [comment]
That line is in both and the working and non-working examples. I intentionally hard coded it to the old value to rule out whether or not the issue was the URL being constructed incorrectly.

**Comment by jrasm91** [comment]
> That line is in both and the working and non-working examples. I intentionally hard coded it to the old value to rule out whether or not the issue was the URL being constructed incorrectly.

Ah, I misunderstood the issue at first. I think this is related to svelte environment variables, which I don't think you are supposed to get from `process.env.*`, but instead from this special store:

https://kit.svelte.dev/docs/modules#$env-static-public

**Comment by bo0tzz** [comment]
That's correct, but I think you have to use $env/dynamic instead https://kit.svelte.dev/docs/modules#$env-dynamic-public

**Comment by ViViDboarder** [comment]
That's it! Thanks! I'm still very confused as to how it fails yet doesn't error though. Must be something with how svelte works I guess.

**Comment by ViViDboarder** [comment]
Ok broke up the changes to a commit for each service and then the examples. I verified this now works when changing the names of each service in the compose file and then updating my `.env`.

File: nginx/start.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#! /bin/bash
+set -e
+
+export IMMICH_WEB_HOST=${IMMICH_WEB_HOST:-immich-web}
```
Comment (alextran1502) [inline nginx/start.sh:4]: Does this syntax means that if there is no env value of `IMMICH_WEB_HOST` the value `immich-web` will be used instead? So let's say a user not updating their `docker-compose.yml` shouldn't have any problem accessing, correct?

File: nginx/start.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+export IMMICH_WEB_PORT=${IMMICH_WEB_PORT:-3000}
+export IMMICH_SERVER_HOST=${IMMICH_SERVER_HOST:-immich-server}
+export IMMICH_SERVER_PORT=${IMMICH_SERVER_PORT:-3001}
+
+envsubst '$IMMICH_WEB_HOST $IMMICH_WEB_PORT $IMMICH_SERVER_HOST $IMMICH_SERVER_PORT' < /etc/nginx/nginx.conf.template > /etc/nginx/nginx.conf
```
Comment (jrasm91) [inline nginx/start.sh:9]: Worth adding a line at the beginning saying that the file is generated and will be overridden on start up?

**Comment by jrasm91** [comment]
I've thought about this a little bit more and if the ultimate goal is to be able to easily swap these out, we might be better off just having an env variable for the URL instead of separate host/port. Something like this:

```
IMMICH_MACHINE_LEARNING_URL=http://immich-machine-learning:3002
IMMICH_SERVER_URL=http://immich-server:2038
```

If you actually want to swap out different machine learning models, etc, it might be worthwhile to have different env per endpoint, so something like:
```
IMMICH_ML_TAG_URL=http://immich-machine-learning:3002/image-classifier/tag-image
IMMICH_ML_OBJ_URL=http://immich-machine-learning:3002/object-detection/detect-object

Basically, I like the idea of making these configurable, but not I'm thinking of consolidating host/port combos into single url values instead. Thoughts?

**Comment by bo0tzz** [comment]
I think configuring the full URL makes sense, there's no need to configure the ports separately. For ML I don't think we need to configure the different endpoints separately - basically that entire setup is mostly a placeholder right now anyways.

**Comment by alextran1502** [comment]
> I've thought about this a little bit more and if the ultimate goal is to be able to easily swap these out, we might be better off just having an env variable for the URL instead of separate host/port. Something like this:
> 
> ```
> IMMICH_MACHINE_LEARNING_URL=http://immich-machine-learning:3002
> IMMICH_SERVER_URL=http://immich-server:2038
> ```
> 
> If you actually want to swap out different machine learning models, etc, it might be worthwhile to have different env per endpoint, so something like:
> 
> ```
> IMMICH_ML_TAG_URL=http://immich-machine-learning:3002/image-classifier/tag-image
> IMMICH_ML_OBJ_URL=http://immich-machine-learning:3002/object-detection/detect-object
> ```
> 
> Basically, I like the idea of making these configurable, but I'm now thinking it'd be even better to consolidate the host/port combos into single url values instead. Thoughts?

I like this approach

File: nginx/start.sh (hunk 1 above)
Comment (ViViDboarder) [reply nginx/start.sh:4]: Yea, exactly. That sets the default should it not be defined or defined as empty.

**Comment by ViViDboarder** [comment]
Switched to URLs. I suppose it's also nice to be able to adjust the scheme as well if you were using TLS between the instances.

File: nginx/nginx.conf [hunk 3]
```diff
@@ -62,7 +61,7 @@ http {
 
       rewrite /api/(.*) /$1 break;
 
-      proxy_pass http://immich-server:3001;
+      proxy_pass $IMMICH_SERVER_URL
```
Comment (jrasm91) [inline nginx/nginx.conf:64]: Is this missing a semi?

File: docker/docker-compose.dev.yml [hunk 4]
```diff
@@ -97,9 +100,13 @@ services:
     ports:
       - 5432:5432
 
-  immich-proxy:
+  immich-proxy-new:
```
Comment (jrasm91) [inline docker/docker-compose.dev.yml:103]: Is this on purpose?

**Review by jrasm91 — APPROVED** [review]
Sweet, looks great. Just a missing semi I think, otherwise lgtm.

**Comment by ViViDboarder** [comment]
Thanks. Resolved the feedback.

File: nginx/Dockerfile [hunk 5]
```diff
@@ -3,6 +3,7 @@ FROM docker.io/nginxinc/nginx-unprivileged:latest
 COPY LICENSE /licenses/LICENSE.txt
 COPY LICENSE /LICENSE
 
-COPY nginx.conf "/etc/nginx/nginx.conf"
+COPY nginx.conf "/etc/nginx/nginx.conf.template"
+COPY start.sh /start.sh
 
-CMD nginx -g "daemon off;"
```
Comment (alextran1502) [inline nginx/Dockerfile:8]: One last question, what would happen if you removed this line?
